### PR TITLE
[luci/pass] Introduce SubstituteExpandDimsToReshapePass

### PR DESCRIPTION
**Pull request by dayo09**
Let's introduces a new luci pass to substitute CircleExpandDims to CircleReshape.

---

Full draft: #14164

### CODE REVIEW

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  return count;
+}
+
+bool substitute_expand_dims_to_reshape(luci::CircleNode *node)
+{
+  auto target_node = dynamic_cast<luci::CircleExpandDims *>(node);
+  if (target_node == nullptr)
+    return false;
+  if (target_node->shape_status() != luci::ShapeStatus::VALID) //
+    return false;
```
Comment (jinevening) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:55]: This check seems not necessary. We typically assume shapes of nodes are valid when a pass is invoked.

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  auto input_node = loco::must_cast<luci::CircleNode *>(target_node->input());
+  if (input_node->rank() <= 0)
+    return false;
+  if (input_node->shape_status() != luci::ShapeStatus::VALID) //
+    return false;
```
Comment (jinevening) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:60]: ditto

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  auto axis_node = loco::must_cast<luci::CircleConst *>(target_node->axis());
```
Comment (jinevening) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:61]: There is no restriction that axis should be a constant.

```suggestion
  auto axis_node = dynamic_cast<luci::CircleConst *>(target_node->axis());
```

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  if (axis_node == nullptr)
+    return false;
+
+  auto axis = axis_node->at<loco::DataType::S32>(0);
```
Comment (jinevening) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:65]: This assumes axis_node's dtype is S32. Could you add a check for that? Or you can make a more general function to support S64.

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  for (int32_t i = 0; i < static_cast<int32_t>(input_node->rank()) + 1; i++)
+  {
+    if (i == axis)
+    {
+      const_node->at<loco::DataType::S32>(i) = 1;
+    }
+    else if (i < axis)
+    {
+      const_node->at<loco::DataType::S32>(i) =
+        input_node->dim(i).known() ? input_node->dim(i).value() : -1;
```
Comment (jinevening) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:93]: It seems that `known()` is already checked before.

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    }
+    else
+    {
+      const_node->at<loco::DataType::S32>(i) =
+        input_node->dim(i - 1).known() ? input_node->dim(i - 1).value() : -1;
+    }
+  }
+  const_node->name(name + "/Reshape/shape");
```
Comment (jinevening) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:101]:
```suggestion
  const_node->name(axis_node->name() + "/Reshape/shape");
```

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+#include <luci/IR/CircleNodes.h>
+
+#include <gtest/gtest.h>
+
+namespace
+{
+
+using uilist = std::initializer_list<uint32_t>;
+using ilist = std::initializer_list<int32_t>;
```
Comment (jinevening) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp:27]: Seems unused.

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/SubstituteExpandDimsToReshapePass.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+#include <bitset>
```
Comment (jinevening) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:22]: What is this header for?

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+  ASSERT_EQ(nullptr, expand_dims);
+  auto reshape_shape = loco::must_cast<luci::CircleConst *>(reshape->shape());
+  ASSERT_EQ(4, reshape_shape->size<loco::DataType::S32>());
+  ASSERT_EQ(16, reshape_shape->at<loco::DataType::S32>(0));
+  ASSERT_EQ(3, reshape_shape->at<loco::DataType::S32>(1));
+  ASSERT_EQ(1, reshape_shape->at<loco::DataType::S32>(2));
+  ASSERT_EQ(1, reshape_shape->at<loco::DataType::S32>(3));
+}
+
+TEST_F(SubstituteExpandDimsToReshapeTest, nothing_to_expand_dims)
```
Comment (jinevening) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp:191]: Could you change this to a negative test and add more negative tests so that the number of negative tests == non-negative tests?

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 1 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:55]: Thanks for the information! ;-D

Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp]: I force-pushed it for some reason, let me mark here the applied changes 😓 

✔️ marked!

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 2 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp]: ✔️ marked!

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 3 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp]: ✔️ marked!

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 4 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp]: 'value_from_circle_const' function, ✔️ marked!

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 5 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp]: Removed, ✔️ marked!

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp (hunk 9 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp]: I added some. :-D

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp (hunk 7 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp]: Removed, ✔️ marked!

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 8 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp]: Removed, ✔️ marked!

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+  ASSERT_EQ(2, reshape_shape->at<loco::DataType::S32>(0));
+  ASSERT_EQ(3, reshape_shape->at<loco::DataType::S32>(1));
+  ASSERT_EQ(4, reshape_shape->at<loco::DataType::S32>(2));
+  ASSERT_EQ(1, reshape_shape->at<loco::DataType::S32>(3));
+}
+
+TEST_F(SubstituteExpandDimsToReshapeTest, simple_with_expand_dims_2)
+{
+  // NOTE Incorrect output shape is allowed
+  _graph.init({16, 3, 1}, {16, 3, 4, 1}, 2);
```
Comment (dayo09) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp:223]: @jinevening This is unexpected from my view. I tried to turn this into negative test by giving it an incorrect Output shape. This case, correct shape whill be ({16,3,1}, {16,3,1,1,}, 2). Do you have any idea why it runs without failure?

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+ * @brief Return value in position on CircleConst with int32 format.
+ */
+int32_t value_from_circle_const(const luci::CircleConst *node, uint32_t idx)
+{
+  assert(node->rank() == 1 && node->dim(0).value() > idx);
+  assert(node->dtype() == loco::DataType::S64 || node->dtype() == loco::DataType::S32);
+
+  if (node->dtype() == loco::DataType::S64)
+    return node->at<loco::DataType::S64>(idx);
+  return static_cast<int32_t>(node->at<loco::DataType::S32>(idx));
```
Comment (seanshpark) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:57]: do we need to use `static_cast<int32_t>` ?

Comment (seanshpark) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:57]: maybe ?
```suggestion
    return static_cast<int32_t>(node->at<loco::DataType::S64>(idx));
  return node->at<loco::DataType::S32>(idx);
```

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+  if (target_node == nullptr)
+    return false;
+  auto input_node = loco::must_cast<luci::CircleNode *>(target_node->input());
+  if (input_node->rank() <= 0)
+    return false;
+  auto axis_node = dynamic_cast<luci::CircleConst *>(target_node->axis());
+  if (axis_node == nullptr)
+    return false;
+
+  int32_t axis = value_from_circle_const(axis_node, 0);
```
Comment (seanshpark) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:72]: normal models will have S32/S64 dtypes. maybe add dtype checks before calling `value_from_circle_const()`

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+  if (axis < 0)
+    axis = axis + static_cast<int32_t>(input_node->rank()) + 1;
```
Comment (seanshpark) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:74]: why +1 ?

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+} // namespace
+
+namespace luci
+{
+
+/**
+ * BEFORE
+ *           |
+ *      [CircleNode]   [CircleConst]
```
Comment (seanshpark) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:120]: there was explanation starting with `@brief Convert expand_dims ... ` above.
it would be better to merge these text graph to above explanation.

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+  SubstituteExpandDimsToReshapeTest() = default;
+
+  void run_pass(void)
+  {
+    while (_shapeinf.run(_graph.g()) || _pass.run(_graph.g()))
+      ;
+  }
+
+protected:
+  ExpandDimsWithConstAxisGraph _graph; // *W*ith *C*onst axis
```
Comment (seanshpark) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp:154]: why use `*W*`, `*C*` ? this is just source code not markdown file.

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+  SubstituteExpandDimsToReshapeNegTest() = default;
+
+  void run_pass(void)
+  {
+    while (_shapeinf.run(_graph.g()) || _pass.run(_graph.g()))
+      ;
+  }
+
+protected:
+  ExpandDimsWithInputAxisGraph _graph; // *W*ith *N*on *C*onst axis
```
Comment (seanshpark) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp:171]: ditto

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+TEST_F(SubstituteExpandDimsToReshapeTest, neg_invalid_axis)
+{
+  _graph.init({16, 3, 1}, {16, 3, 1, 1}, 5);
+
+  EXPECT_DEATH(run_pass(), "assert");
+}
+
+TEST_F(SubstituteExpandDimsToReshapeNegTest, neg_non_const_axis)
```
Comment (seanshpark) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp:246]: plz add negative tests name ending with `_NEG` with count(positive) <= count(negative)

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp (hunk 15 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp:154]: Before I named it with abbreviations. Now it's renamed and I forgot to remove them 😅. Let me remove them :-D

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 13 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:74]: Example) 
input node: {1, 3}
input node's rank: 2
axis: -1

axis should point to the last index of **new** input node.
expected new input node: {1, 3, 1}
expected last index: 2

Likewise, axis must be the current last index + 1. Therefore (input_node->rank() + 1).

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 12 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:72]: Do you mean something different from below? https://github.com/Samsung/ONE/blob/df6b80399e2f873335f797e4513ccc7c7a264e8d/compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp#L53

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 11 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:57]: Let me remove it , thanks :-D

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 12 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:72]: Aha, I got it. Let me add some phrase like 

```cpp
  if (node->dtype() != loco::DataType::S64 && node->dtype() != loco::DataType::S32){
    // Unexpected type of model
    return false;
  }
```

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 14 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:120]: @seanshpark Sure. :-D

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 13 above)
Comment (seanshpark) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:98]: as `+1` is not like other Ops axis with -1, it would be nice to have a comment explaining this in the source.

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 12 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:96]: @seanshpark Or maybe assert is rather proper?

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+
+  for (uint32_t i = 0; i < node->rank(); ++i)
+    if (!node->dim(i).known())
+      ++count;
+
+  return count;
+}
+
+/**
+ * @brief Return value in position on CircleConst with int32 format.
```
Comment (jinevening) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:66]:
```suggestion
 * @brief Return value in position on CircleConst with int32/int64 format.
```

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+    else if (i < axis)
+    {
+      const_node->at<loco::DataType::S32>(i) = input_node->dim(i).value();
+    }
+    else
+    {
+      const_node->at<loco::DataType::S32>(i) = input_node->dim(i - 1).value();
+    }
+  }
+  const_node->name(name + "/Reshape/shape");
```
Comment (jinevening) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:133]: Could you add origin to const_node?

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp (hunk 10 above)
Comment (jinevening) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp]: That's because no one checks the shape is correct in the test code.

Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp]: @jinevening Even shape inference code doen't check?

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 18 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:66]: @jinevening It was to mean that it **returns** int32 value, so let me clarify that :-D

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 19 above)
Comment (dayo09) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:133]: Does it help that const node having origin, too? 
I took a reference from SubstitutePackToReshapePass, 
- [ ]if so, let me add some code there, too.

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp (hunk 10 above)
Comment (jinevening) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp]: Shape inference only checks if inputs have invalid shapes fo inferring output shape. So, if inputs had valid shapes, the output shape would be inferred and overwritten even though the previous output shape was wrong.

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 19 above)
Comment (jinevening) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:135]: > Does it help that const node having origin, too?

Yes. It helps to identiy the node name when we do profiling in the backend.

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 18 above)
Comment (jinevening) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp]: Thanks!

**Review by jinevening — APPROVED** [review]
LGTM Thank you

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp (hunk 12 above)
Comment (seanshpark) [reply compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:96]: assert in the method is OK to confirm the input arguments.
conditional check and return false here looks OK.

File: compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+
+  if (axis_node->dtype() != loco::DataType::S64 && axis_node->dtype() != loco::DataType::S32)
+  {
+    // Abnormal model with unexpected dtype in axis
+    return false;
+  }
+
+  int32_t axis = value_from_circle_const(axis_node, 0);
+  if (axis < 0)
+    // WHY ADD +1?
```
Comment (seanshpark) [inline compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp:98]: 👍🏻

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

**Comment by seanshpark** [comment]
@hseok-oh , can you check our CI status? `nncc-release` / `tizen-gbs` doesn't seem to be running ...

**Comment by hseok-oh** [comment]
@nnfw-bot nncc-release tizen-gbs

**Comment by dayo09** [comment]
@nnfw-bot nncc-release tizen-gbs

**Comment by dayo09** [comment]
Let me rebase it to see if jenkins workers are activated..

**Comment by dayo09** [comment]
@seanshpark Finally, tests are all done! PTAL. Thanks! :-D
